File: frontend/src/utils/teifGenerator.js
/**
 * TEIF V1.8.8 XML GENERATOR - El Fatoora PFE
 * Conforme au schéma facture_INVOIC_V1.8.8_withoutSig.xsd (Tunisie TradeNet)
 */

export const STAMP_DUTY = 1.000;

/**
 * Normalise le Matricule Fiscal à exactement 13 caractères
 * Format attendu: 7chiffres + lettre[ABDNP] + lettre[CMNP] + 3chiffres
 */
const normaliseMatricule = (mf) => {
  let clean = (mf || '').replace(/[^a-zA-Z0-9]/g, '').toUpperCase();
  if (clean.length < 13) clean = clean.padEnd(13, '0');
  if (clean.length > 13) clean = clean.substring(0, 13);
  return clean;
};

/**
 * Formate une date en ddMMyy (format exigé par le XSD DtmDetailType)
 * Ex: 2026-05-04 → 040526
 */
const formatDateDDMMYY = (dateStr) => {
  const d = new Date(dateStr || new Date());
  const dd = String(d.getDate()).padStart(2, '0');
  const mm = String(d.getMonth() + 1).padStart(2, '0');
  const yy = String(d.getFullYear()).slice(-2);
  return `${dd}${mm}${yy}`;
};

export const generateTeifXml = (issuer, invoice) => {
  const senderMF   = normaliseMatricule(issuer?.matriculeFiscal || issuer?.matricule || '');
  const receiverMF = normaliseMatricule(invoice.clientMatricule);

  const rawItems = invoice.lines || invoice.items || [];
  const items = rawItems.length > 0
    ? rawItems
    : [{ description: 'Ligne de test', qty: 1, puht: 0, unitPriceHT: 0, tvaRate: 19 }];

  // Grouper les lignes par taux TVA
  const tvaGroups = {};
  items.forEach(item => {
    const rate = parseFloat(item.tvaRate || item.tva || 19);
    const pu   = parseFloat(item.puht || item.unitPriceHT || 0);
    const qty  = parseFloat(item.qty || 0);
    const ht   = qty * pu;
    if (!tvaGroups[rate]) tvaGroups[rate] = { ht: 0, tva: 0 };
    tvaGroups[rate].ht  += ht;
    tvaGroups[rate].tva += ht * (rate / 100);
  });

  const totalHT   = parseFloat(invoice.totalHT  || invoice.totals?.ht  || 0).toFixed(3);
  const totalTVA  = parseFloat(invoice.totalTVA || invoice.totals?.tva || 0).toFixed(3);
  const totalTTC  = parseFloat(invoice.totalTTC || invoice.totals?.ttc || 0).toFixed(3);
  const stampDuty = parseFloat(invoice.stampDuty || STAMP_DUTY).toFixed(3);
  const dateStr   = formatDateDDMMYY(invoice.date);
  const docId     = invoice.number || invoice.invoiceNumber || 'FAC-0001';
  const senderName   = issuer?.entreprise || issuer?.name || 'Émetteur';
  const senderAddr   = issuer?.address || 'Tunis, Tunisie';
  const receiverName = invoice.clientName || 'Client';
  const receiverAddr = invoice.clientAddress || 'Adresse Client';

  return `<?xml version="1.0" encoding="UTF-8"?>
<TEIF version="1.8.8" controlingAgency="TTN" xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
  <InvoiceHeader>
    <MessageSenderIdentifier type="I-01">${senderMF}</MessageSenderIdentifier>
    <MessageRecieverIdentifier type="I-01">${receiverMF}</MessageRecieverIdentifier>
  </InvoiceHeader>
  <InvoiceBody>
    <Bgm>
      <DocumentIdentifier>${docId}</DocumentIdentifier>
      <DocumentType code="I-11">Facture</DocumentType>
    </Bgm>
    <Dtm>
      <DateText functionCode="I-31" format="ddMMyy">${dateStr}</DateText>
    </Dtm>

    <PartnerSection>
      <PartnerDetails functionCode="I-61">
        <Nad>
          <PartnerIdentifier type="I-01">${senderMF}</PartnerIdentifier>
          <PartnerName nameType="Qualification">${senderName}</PartnerName>
          <PartnerAdresses>
            <AdressDescription>${senderAddr}</AdressDescription>
            <Country codeList="ISO_3166-1">TN</Country>
          </PartnerAdresses>
        </Nad>
      </PartnerDetails>
      <PartnerDetails functionCode="I-62">
        <Nad>
          <PartnerIdentifier type="I-01">${receiverMF}</PartnerIdentifier>
          <PartnerName nameType="Qualification">${receiverName}</PartnerName>
          <PartnerAdresses>
            <AdressDescription>${receiverAddr}</AdressDescription>
            <Country codeList="ISO_3166-1">TN</Country>
          </PartnerAdresses>
        </Nad>
      </PartnerDetails>
    </PartnerSection>

    <LinSection>
${items.map((item, idx) => {
  const pu  = parseFloat(item.puht || item.unitPriceHT || 0);
  const qty = parseFloat(item.qty || 0);
  const ht  = (qty * pu).toFixed(3);
  const rate = parseInt(item.tvaRate || item.tva || 19);
  const itemId = item.id > 0 ? item.id : idx + 1;
  return `      <Lin>
        <ItemIdentifier>${idx + 1}</ItemIdentifier>
        <LinImd lang="fr">
          <ItemCode>${itemId}</ItemCode>
          <ItemDescription>${item.description || ''}</ItemDescription>
        </LinImd>
        <LinQty>
          <Quantity measurementUnit="UN">${qty.toFixed(3)}</Quantity>
        </LinQty>
        <LinTax>
          <TaxTypeName code="I-1602">TVA</TaxTypeName>
          <TaxDetails>
            <TaxRate>${rate}</TaxRate>
          </TaxDetails>
        </LinTax>
        <LinMoa>
          <MoaDetails>
            <Moa currencyCodeList="ISO_4217" amountTypeCode="I-171">
              <Amount currencyIdentifier="TND">${ht}</Amount>
            </Moa>
          </MoaDetails>
        </LinMoa>
      </Lin>`;
}).join('\n')}
    </LinSection>

    <InvoiceMoa>
      <AmountDetails>
        <MoaDetails>
          <Moa currencyCodeList="ISO_4217" amountTypeCode="I-179">
            <Amount currencyIdentifier="TND">${totalHT}</Amount>
          </Moa>
        </MoaDetails>
      </AmountDetails>
      <AmountDetails>
        <MoaDetails>
          <Moa currencyCodeList="ISO_4217" amountTypeCode="I-176">
            <Amount currencyIdentifier="TND">${totalTVA}</Amount>
          </Moa>
        </MoaDetails>
      </AmountDetails>
      <AmountDetails>
        <MoaDetails>
          <Moa currencyCodeList="ISO_4217" amountTypeCode="I-175">
            <Amount currencyIdentifier="TND">${stampDuty}</Amount>
          </Moa>
        </MoaDetails>
      </AmountDetails>
      <AmountDetails>
        <MoaDetails>
          <Moa currencyCodeList="ISO_4217" amountTypeCode="I-177">
            <Amount currencyIdentifier="TND">${totalTTC}</Amount>
          </Moa>
        </MoaDetails>
      </AmountDetails>
    </InvoiceMoa>

    <InvoiceTax>
${Object.entries(tvaGroups).map(([rate, vals]) => `      <InvoiceTaxDetails>
        <Tax>
          <TaxTypeName code="I-1602">TVA</TaxTypeName>
          <TaxDetails>
            <TaxRate>${parseInt(rate)}</TaxRate>
          </TaxDetails>
        </Tax>
        <AmountDetails>
          <MoaDetails>
            <Moa currencyCodeList="ISO_4217" amountTypeCode="I-179">
              <Amount currencyIdentifier="TND">${parseFloat(vals.ht).toFixed(3)}</Amount>
            </Moa>
          </MoaDetails>
        </AmountDetails>
        <AmountDetails>
          <MoaDetails>
            <Moa currencyCodeList="ISO_4217" amountTypeCode="I-176">
              <Amount currencyIdentifier="TND">${parseFloat(vals.tva).toFixed(3)}</Amount>
            </Moa>
          </MoaDetails>
        </AmountDetails>
      </InvoiceTaxDetails>`).join('\n')}
      <InvoiceTaxDetails>
        <Tax>
          <TaxTypeName code="I-1601">Droit de timbre</TaxTypeName>
          <TaxDetails>
            <TaxRate>0</TaxRate>
          </TaxDetails>
        </Tax>
        <AmountDetails>
          <MoaDetails>
            <Moa currencyCodeList="ISO_4217" amountTypeCode="I-175">
              <Amount currencyIdentifier="TND">${stampDuty}</Amount>
            </Moa>
          </MoaDetails>
        </AmountDetails>
      </InvoiceTaxDetails>
    </InvoiceTax>
  </InvoiceBody>
  <ds:Signature/>
</TEIF>`;
};

export const downloadXml = (xml, filename = 'facture.xml') => {
  const blob = new Blob([xml], { type: 'application/xml' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  a.click();
};

File: backend/Utils/TeifGenerator.cs
using System.Text.RegularExpressions;
using System.Xml.Linq;
using backend.Models;

namespace backend.Utils
{
    public static class TeifGenerator
    {
        // TEIF v1.8.8 has NO targetNamespace — elements are unqualified
        private static readonly XNamespace ds = "http://www.w3.org/2000/09/xmldsig#";

        // -------------------------------------------------------------------
        // Helper: normalise the Matricule Fiscal to exactly 13 chars
        // Format: 7 digits + letter + letter + letter + 3 digits  (I-01)
        // -------------------------------------------------------------------
        private static string NormaliseMatricule(string mf)
        {
            string clean = Regex.Replace(mf ?? "", "[^a-zA-Z0-9]", "").ToUpper();
            // Pad or truncate to 13 characters so the XSD assert passes
            if (clean.Length < 13)
                clean = clean.PadRight(13, '0');
            else if (clean.Length > 13)
                clean = clean.Substring(0, 13);
            return clean;
        }

        // -------------------------------------------------------------------
        // Helper: format a date as ddMMyy (required by XSD DtmDetailType)
        // -------------------------------------------------------------------
        private static string FormatDate(DateTime d) => d.ToString("ddMMyy");

        // -------------------------------------------------------------------
        // Helper: build a <Moa> element with the mandatory attributes
        //   amountTypeCode codes (v1.8.8):
        //     I-171 … I-188  — see XSD enumeration
        //   Typical usage:
        //     I-171  Line net amount (HT per line)
        //     I-176  Tax amount (TVA total)
        //     I-177  Total TTC
        //     I-179  Total HT (taxable base)
        //     I-175  Stamp duty
        // -------------------------------------------------------------------
        private static XElement MoaElement(string amountTypeCode, decimal amount)
        {
            return new XElement("Moa",
                new XAttribute("currencyCodeList", "ISO_4217"),
                new XAttribute("amountTypeCode", amountTypeCode),
                new XElement("Amount",
                    new XAttribute("currencyIdentifier", "TND"),
                    amount.ToString("F3")
                )
            );
        }

        // -------------------------------------------------------------------
        // Helper: wrap a Moa inside an AmountDetails container
        // -------------------------------------------------------------------
        private static XElement AmountDetails(string amountTypeCode, decimal amount)
        {
            return new XElement("AmountDetails",
                new XElement("MoaDetails",
                    MoaElement(amountTypeCode, amount)
                )
            );
        }

        // -------------------------------------------------------------------
        // Main entry point
        // -------------------------------------------------------------------
        public static string GenerateXml(Invoice invoice, Company company)
        {
            string senderMF   = NormaliseMatricule(company.RegistrationNumber);
            string receiverMF = NormaliseMatricule(invoice.ClientMatricule);

            var lines = invoice.Lines ?? new List<InvoiceLine>();

            var doc = new XDocument(
                new XDeclaration("1.0", "UTF-8", null),
                new XElement("TEIF",
                    new XAttribute("version", "1.8.8"),
                    new XAttribute("controlingAgency", "TTN"),
                    new XAttribute(XNamespace.Xmlns + "ds", ds.NamespaceName),

                    // ── InvoiceHeader ─────────────────────────────────────
                    new XElement("InvoiceHeader",
                        new XElement("MessageSenderIdentifier",
                            new XAttribute("type", "I-01"),
                            senderMF
                        ),
                        new XElement("MessageRecieverIdentifier",
                            new XAttribute("type", "I-01"),
                            receiverMF
                        )
                    ),

                    // ── InvoiceBody ───────────────────────────────────────
                    new XElement("InvoiceBody",

                        // Bgm — document identification
                        new XElement("Bgm",
                            new XElement("DocumentIdentifier", invoice.InvoiceNumber ?? invoice.Id.ToString()),
                            new XElement("DocumentType",
                                new XAttribute("code", "I-11"),   // I-11 = Facture
                                "Facture"
                            )
                        ),

                        // Dtm — invoice date  (functionCode I-31 = Invoice date)
                        new XElement("Dtm",
                            new XElement("DateText",
                                new XAttribute("functionCode", "I-31"),
                                new XAttribute("format", "ddMMyy"),
                                FormatDate(invoice.Date)
                            )
                        ),

                        // PartnerSection — Seller (I-61) + Buyer (I-62)
                        new XElement("PartnerSection",

                            // Seller
                            new XElement("PartnerDetails",
                                new XAttribute("functionCode", "I-61"),
                                new XElement("Nad",
                                    new XElement("PartnerIdentifier",
                                        new XAttribute("type", "I-01"),
                                        senderMF
                                    ),
                                    new XElement("PartnerName",
                                        new XAttribute("nameType", "Qualification"),
                                        company.Name ?? ""
                                    ),
                                    new XElement("PartnerAdresses",
                                        new XElement("AdressDescription", company.Address ?? ""),
                                        new XElement("Country",
                                            new XAttribute("codeList", "ISO_3166-1"),
                                            "TN"
                                        )
                                    )
                                )
                            ),

                            // Buyer
                            new XElement("PartnerDetails",
                                new XAttribute("functionCode", "I-62"),
                                new XElement("Nad",
                                    new XElement("PartnerIdentifier",
                                        new XAttribute("type", "I-01"),
                                        receiverMF
                                    ),
                                    new XElement("PartnerName",
                                        new XAttribute("nameType", "Qualification"),
                                        invoice.ClientName ?? ""
                                    ),
                                    new XElement("PartnerAdresses",
                                        new XElement("AdressDescription", invoice.ClientAddress ?? ""),
                                        new XElement("Country",
                                            new XAttribute("codeList", "ISO_3166-1"),
                                            "TN"
                                        )
                                    )
                                )
                            )
                        ),

                        // LinSection — one <Lin> per invoice line
                        new XElement("LinSection",
                            lines.Select((line, idx) =>
                                new XElement("Lin",
                                    new XElement("ItemIdentifier", (idx + 1).ToString()),
                                    new XElement("LinImd",
                                        new XAttribute("lang", "fr"),
                                        new XElement("ItemCode", line.Id > 0 ? line.Id.ToString() : (idx + 1).ToString()),
                                        new XElement("ItemDescription", line.Description ?? "")
                                    ),
                                    new XElement("LinQty",
                                        new XElement("Quantity",
                                            new XAttribute("measurementUnit", "UN"),
                                            line.Qty.ToString("F3")
                                        )
                                    ),
                                    new XElement("LinTax",
                                        new XElement("TaxTypeName",
                                            new XAttribute("code", "I-1602"),  // I-1602 = TVA
                                            "TVA"
                                        ),
                                        new XElement("TaxDetails",
                                            new XElement("TaxRate", ((int)line.TvaRate).ToString())
                                        )
                                    ),
                                    new XElement("LinMoa",
                                        // I-171 = line net amount (HT)
                                        new XElement("MoaDetails",
                                            MoaElement("I-171", line.TotalHT)
                                        )
                                    )
                                )
                            )
                        ),

                        // InvoiceMoa — invoice-level amounts
                        new XElement("InvoiceMoa",
                            // I-179 = Total taxable amount (HT)
                            new XElement("AmountDetails",
                                new XElement("MoaDetails", MoaElement("I-179", invoice.TotalHT))
                            ),
                            // I-176 = Total tax amount (TVA)
                            new XElement("AmountDetails",
                                new XElement("MoaDetails", MoaElement("I-176", invoice.TotalTVA))
                            ),
                            // I-175 = Stamp duty
                            new XElement("AmountDetails",
                                new XElement("MoaDetails", MoaElement("I-175", invoice.StampDuty))
                            ),
                            // I-177 = Total payable (TTC)
                            new XElement("AmountDetails",
                                new XElement("MoaDetails", MoaElement("I-177", invoice.TotalTTC))
                            )
                        ),

                        // InvoiceTax — one entry per TVA rate group + stamp duty
                        new XElement("InvoiceTax",
                            // TVA groups
                            lines.GroupBy(l => l.TvaRate).Select(g =>
                                new XElement("InvoiceTaxDetails",
                                    new XElement("Tax",
                                        new XElement("TaxTypeName",
                                            new XAttribute("code", "I-1602"),
                                            "TVA"
                                        ),
                                        new XElement("TaxDetails",
                                            new XElement("TaxRate", ((int)g.Key).ToString())
                                        )
                                    ),
                                    // taxable base for this rate
                                    new XElement("AmountDetails",
                                        new XElement("MoaDetails",
                                            MoaElement("I-179", g.Sum(l => l.TotalHT))
                                        )
                                    ),
                                    // tax amount for this rate
                                    new XElement("AmountDetails",
                                        new XElement("MoaDetails",
                                            MoaElement("I-176", g.Sum(l => l.TotalTVA))
                                        )
                                    )
                                )
                            ),
                            // Stamp duty (I-1601)
                            new XElement("InvoiceTaxDetails",
                                new XElement("Tax",
                                    new XElement("TaxTypeName",
                                        new XAttribute("code", "I-1601"),
                                        "Droit de timbre"
                                    ),
                                    new XElement("TaxDetails",
                                        new XElement("TaxRate", "0")
                                    )
                                ),
                                new XElement("AmountDetails",
                                    new XElement("MoaDetails",
                                        MoaElement("I-175", invoice.StampDuty)
                                    )
                                )
                            )
                        )
                    ),

                    // ── ds:Signature placeholder (filled by signing service) ──
                    new XElement(ds + "Signature")
                )
            );

            return doc.ToString();
        }
    }
}
